### PR TITLE
fix(python)!: raise IndexError on out of bounds series index

### LINKED ISSUE #9877: Invalid indexes of a Series should raise `IndexError`
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

data = []

pl.Series(data)[0]  # exceptions.ComputeError: index 0 is out of bounds for sequence of size 0
pl.Series(data)[-1]  # OverflowError: can't convert negative int to unsigned

data[0]  # IndexError: list index out of range
data[-1]  # IndexError: list index out of range
```

### Issue description

Since we allow accessing values by indexing, we should raise errors that are consistent with what Python users would expect.

### Expected behavior

An `IndexError` should be raised in all cases where the index is out of range.

### Installed versions

master

### PR DESCRIPTION
**Pull request by tjkuson**
Closes #9877

### CODE REVIEW

**Review by alexander-beedie — CHANGES_REQUESTED** [review]
Nice; makes perfect sense that we should do this. Can you enrich the error slightly? Something like the following would provide additional useful pointers:

`IndexError(f"Series index {item} is out of bounds (len={self.len()})")`

**Comment by tjkuson** [comment]
> Technically this is a breaking change as the `Exception` type is going to be different, so have tagged as such. Otherwise, this is looking good to me - thanks @tjkuson :)

Ah, sorry I missed that!

**Review by stinodego — CHANGES_REQUESTED** [review]
Thanks for the PR! However, I believe this is not the way to go.

You're checking explicitly for bounds, while this is already done in the `get_idx` method on the Rust side.

I think we should update the Rust side to throw a less generic error (`ComputeError` can be anything...), and then make the Rust binding emit an IndexError. The OverflowError happens before entering the Rust function - you can use try/except to catch it and convert to an IndexError.

**Comment by ritchie46** [comment]
Good point @stinodego. I think we should add an `PolarsError::OutOfBounds`, or make `PolarsError::ShapeMismatch` more generic.

File: py-polars/polars/series/series.py [hunk 1]
```diff
@@ -974,8 +974,16 @@ def __getitem__(
...
+            if item > self.len() - 1:
+                raise IndexError(
+                    f"Series index {item} is out of bounds (len={self.len()})"
+                )
             if item < 0:
                 item = self.len() + item
+                if item < 0:
+                    raise IndexError(
+                        f"Series index {item - self.len()} is out of bounds (len={self.len()})"
+                    )
```
Comment (AdamBajger) [inline py-polars/polars/series/series.py:986]: How about a simple
```suggestion
           if (item % self.len) >= self.len:
               raise IndexError(
                  f"Series index {item} is out of bounds (len={self.len()})"
               )
```

Comment (AdamBajger) [reply py-polars/polars/series/series.py:986]: I see, haven't read the previous replies. I might as well create my own merge request then...

**Comment by tjkuson** [comment]
I have had a go at adding `PolarsError::OutOfBounds`, is this the preferred approach? Also, the GitHub Actions seems to be failing due to a spurious network error rather than the PR per se.

**Comment by stinodego** [comment]
@tjkuson Thanks for the effort - this is in the right direction but doesn't really get there all the way. I opened https://github.com/pola-rs/polars/pull/11061 to finalize this fix.

**Pull request closed without merging**